File: Hashing/website_cache_hashing.c
/* File Name: website_cache_hashing.c */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SIZE 10
#define EMPTY ""

// Structure for cache entry
struct Cache {
    char url[100];
    char content[200];
};

struct Cache cache[SIZE];

// Hash Function (Simple String Hash)
int hashFunction(char *url) {
    int sum = 0;
    for (int i = 0; url[i] != '\0'; i++)
        sum += url[i];
    return sum % SIZE;
}

// Initialize cache
void initialize() {
    for (int i = 0; i < SIZE; i++)
        strcpy(cache[i].url, EMPTY);
}

// Access Website
void accessWebsite(char *url) {
    int index = hashFunction(url);
    int start = index;

    while (strcmp(cache[index].url, EMPTY) != 0) {
        if (strcmp(cache[index].url, url) == 0) {
            printf("Cache HIT: %s\nContent: %s\n", url, cache[index].content);
            return;
        }
        index = (index + 1) % SIZE;
        if (index == start)
            break;
    }

    printf("Cache MISS: Fetching from server...\n");
    strcpy(cache[index].url, url);
    sprintf(cache[index].content, "HTML data of %s", url);
    printf("Stored in cache.\n");
}

// Display Cache
void display() {
    printf("\nCache Table:\n");
    for (int i = 0; i < SIZE; i++) {
        if (strcmp(cache[i].url, EMPTY) == 0)
            printf("Index %d -> EMPTY\n", i);
        else
            printf("Index %d -> %s\n", i, cache[i].url);
    }
}

int main() {
    initialize();
    int choice;
    char url[100];

    while (1) {
        printf("\n1.Access Website\n2.Display Cache\n3.Exit\n");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter URL: ");
                scanf("%s", url);
                accessWebsite(url);
                break;
            case 2:
                display();
                break;
            case 3:
                exit(0);
        }
    }
}

File: Hashing/website_cache_hashing.py
# File Name: website_cache_hashing.py

class WebsiteCache:
    def __init__(self, size=10):
        self.size = size
        self.cache = [None] * size

    def hash_function(self, url):
        return sum(ord(c) for c in url) % self.size

    def access_website(self, url):
        index = self.hash_function(url)
        start = index

        while self.cache[index] is not None:
            if self.cache[index]["url"] == url:
                print(f"Cache HIT: {url}")
                print(f"Content: {self.cache[index]['content']}")
                return
            index = (index + 1) % self.size
            if index == start:
                break

        print("Cache MISS: Fetching from server...")
        self.cache[index] = {
            "url": url,
            "content": f"HTML data of {url}"
        }
        print("Stored in cache.")

    def display(self):
        print("\nCache Table:")
        for i in range(self.size):
            if self.cache[i] is None:
                print(f"Index {i} -> EMPTY")
            else:
                print(f"Index {i} -> {self.cache[i]['url']}")


if __name__ == "__main__":
    cache = WebsiteCache()

    while True:
        print("\n1.Access Website\n2.Display Cache\n3.Exit")
        choice = int(input("Enter choice: "))

        if choice == 1:
            url = input("Enter URL: ")
            cache.access_website(url)
        elif choice == 2:
            cache.display()
        elif choice == 3:
            break
